Uninitialized variable fix for ISNUMERIC function with varchar and nvarchar arguments

manisha-deshpande opened the pull request:
### Description

This change addresses a bug about the incorrect behavior of the ISNUMERIC function in Babelfish when handling string variables (NVARCHAR and VARCHAR). 

Currently, production Babelfish instances (specifically `PostgreSQL 16.2 on aarch64-unknown-linux-gnu (Babelfish 4.1.1)`) incorrectly handle the ISNUMERIC function for string variables (NVARCHAR and VARCHAR), returning 0 (false) for valid numeric string representations. With this change, ISNUMERIC will correctly return 1 (true) for valid numeric strings, aligning its behavior with SQL Server.

The issue stemmed from an uninitialized variable `typmod` containing garbage values, which impacted the behavior of the `numeric_in` function, invoked by the `varchar2numeric` function.
In the production instances, the typmod variable, happens to point to an incorrect positive garbage values, bypassing the `is_valid_numeric_typmod` check, trickling down into an error. The change adds two arguments to the numeric_in call made in varchar2numeric, namely `OID (Elem Type)` and `TYPMOD`, which are then used to initialize the variables in numeric_in to default values, ensuring that no garbage values are used. It was also observed that numeric_in is invoked similarly with three arguments in other functions such as `cstring_to_numeric`, `fixeddecimal_numeric` etc, except this particular function.

Test cases test `ISNUMERIC` with varchar/nvarchar variables, invoking `varchar2numeric`. Existing test cases `babel_isnumeric`, `babel_isnumeric-vu-verify` already cover various other data types and scenarios: numeric data types and string literals.

### Issues Resolved

BABEL-5129


### Test Scenarios Covered ###
* **Use case based -**

```
-- Test varchar
SELECT ISNUMERIC(varchar_type)
FROM babel_5129
GO
~~START~~
int
1
~~END~~


-- Test varchar variable
DECLARE @v varchar(20);
SELECT @v = '12.3420000000';
SELECT ISNUMERIC(@v), LEN(@v), DATALENGTH(@v)
GO
~~START~~
int#!#int#!#int
1#!#13#!#13
~~END~~
```
```
-- Test nvarchar
SELECT ISNUMERIC(nvarchar_type)
FROM babel_5129
GO
~~START~~
int
1
~~END~~


-- Test nvarchar variable
DECLARE @nv nvarchar(10);
SELECT @nv = '12.3420000000';
SELECT ISNUMERIC(@nv), LEN(@nv), DATALENGTH(@nv)
GO
~~START~~
int#!#int#!#int
1#!#10#!#10
~~END~~

```
* **Existing Test cases -**

- test/JDBC/input/babel_isnumeric.sql
- test/JDBC/input/babel_isnumeric-vu-verify.sql

```
CREATE TABLE test_isnumeric (
    bigint_type bigint,
    int_type int,
    smallint_type smallint,
    tinyint_type tinyint,
    bit_type bit,
    decimal_type decimal(5,2),
    numeric_type numeric(10,5),
    float_type float,
    real_type real,
    money_type money,
    smallmoney_type money
)
GO
...
-- Test bigint
SELECT ISNUMERIC(bigint_type)
FROM test_isnumeric
GO
...
-- Test valid and invalid operators and literals
select isnumeric(' + $ 1.1234')
GO

select isnumeric('abcdefghijklmnop')
GO
...
```

* **Boundary conditions -**

```
-- Test empty varchar variable
DECLARE @v varchar(20);
SELECT @v = '';
SELECT ISNUMERIC(@v), LEN(@v), DATALENGTH(@v)
GO
~~START~~
int#!#int#!#int
0#!#0#!#0
~~END~~


-- Test empty nvarchar variable
DECLARE @nv nvarchar(10);
SELECT @nv = '';
SELECT ISNUMERIC(@nv), LEN(@nv), DATALENGTH(@nv)
GO
~~START~~
int#!#int#!#int
0#!#0#!#0
~~END~~
```

```
-- Test varchar with number argument that exceeds range of bigint.
DECLARE @v varchar(20);
SELECT @v = '9223372036854775807';
SELECT ISNUMERIC(@v), LEN(@v), DATALENGTH(@v)
GO
~~START~~
int#!#int#!#int
1#!#19#!#19
~~END~~


DECLARE @v varchar(20);
SELECT @v = '-9223372036854775808';
SELECT ISNUMERIC(@v), LEN(@v), DATALENGTH(@v)
GO
~~START~~
int#!#int#!#int
1#!#20#!#20
~~END~~
```

```
-- Test nvarchar with lengthy numeric value
DECLARE @nv nvarchar;
SELECT @nv = '12345678901234567890123456789012345';
SELECT ISNUMERIC(@nv), LEN(@nv), DATALENGTH(@nv)
GO
~~START~~
int#!#int#!#int
1#!#1#!#1 --LEN=1 matches SQL server behavior (BABEL-1513)
~~END~~
```

* **Negative test cases -**
```
-- Test varchar variable with invalid numeric
DECLARE @v varchar(20);
SELECT @v = '12.34.20000000';
SELECT ISNUMERIC(@v), LEN(@v), DATALENGTH(@v)
GO
~~START~~
int#!#int#!#int
0#!#14#!#14
~~END~~


-- Test nvarchar variable with invalid numeric
DECLARE @nv nvarchar(10);
SELECT @nv = '12.34.20000000';
SELECT ISNUMERIC(@nv), LEN(@nv), DATALENGTH(@nv)
GO
~~START~~
int#!#int#!#int
0#!#10#!#10
~~END~~

```

* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

manisha-deshpande committed:
Fix uninitialised variable issue related to varchar2numeric function

Task: BABEL-5129
Signed-off-by: manisha-deshpande <mmdeshp@amazon.com>

Commit:
Test cases for isnumeric function with varchar and nvarchar variables

Task: BABEL-5129
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Jakeowen1 commented on contrib/babelfishpg_common/src/varchar.c:
```diff
@@ -922,7 +922,11 @@ varchar2numeric(PG_FUNCTION_ARGS)
-	result = DatumGetNumeric(DirectFunctionCall1(numeric_in, CStringGetDatum(str)));
+	/* Updating numeric_in call with additional expected arguments `Elem type Oid` and `typmod` - BABEL-5129 */
+	result = DatumGetNumeric(DirectFunctionCall3(numeric_in, 
+        CStringGetDatum(str), 
```
nit: please indent the arguments similar to the example below
![Screenshot 2024-09-19 at 12 08 48 PM](https://github.com/user-attachments/assets/d2e0d285-ad4f-4298-bd11-8de3baea5d5e)

Jakeowen1 commented on contrib/babelfishpg_common/src/varchar.c:
```diff
@@ -922,7 +922,11 @@ varchar2numeric(PG_FUNCTION_ARGS)
 
 	str = varchar2cstring(source);
-	result = DatumGetNumeric(DirectFunctionCall1(numeric_in, CStringGetDatum(str)));
+	/* Updating numeric_in call with additional expected arguments `Elem type Oid` and `typmod` - BABEL-5129 */
```
Not sure if this comment is necessary

Jakeowen1 commented on test/JDBC/expected/BABEL-5129-vu-prepare.out:
```diff
@@ -0,0 +1,28 @@
+
+-- Tests for ISNUMERIC function with varchar and nvarchar variables
+CREATE TABLE babel_5129 (
+    int_type int,
```
Please add tests for `bigint, smallint, tinyint, bit, float, real` datatypes, as well as special characters that should return 1 such as `$, +, -`

Jakeowen1 replied:
(If this is already tested in another file please add these examples to the description)

Jakeowen1 commented on test/JDBC/expected/BABEL-5129-vu-verify.out:
```diff
@@ -0,0 +1,131 @@
+
+-- Tests for ISNUMERIC function with varchar and nvarchar variables
```
Are you able to reproduce the faulty behavior in open source before your change was added? If so we need to add some upgrade tests that demonstrate the behavior before/after this change was made. Additionally, please add your test file name to the `JDBC/upgrade/latest/schedule` file

manisha-deshpande replied:
The faulty behavior occurs only in the production environment, where the uninitialized variable `tymod` points to positive garbage values, leading to a handled error and incorrect result. 

This issue is not reproduced in pre-prod and local environments due to negative garbage values, which behave like the default negative value (-1) and do not alter the intended flow. 

GDB debugging confirms that the fix will prevent assignment of unintended positive or negative garbage values, and hence, should fix the issue in production.

Commit:
removing unintended merged commits

Fixing pr comments, added test name to jdbc latest schedule

Task: BABEL-5129
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Fixed issue in SP_TABLES() when @table_name parameter has square brackets around underscore (#2940)

Currently, SP_TABLES procedure does not support wild card [], which matches to any character that is specified in
square bracket. Due to this SP_TABLES procedure returns different output when @table_name parameter value is
table_1 and table[_]1. This issue exists because previously we were using helper function sp_tables_internal whose
body language was plpgsql hence we were using PostgresSQL Like operator for comparison, which does not support
square bracket wild card expression. Fixed the issue by removing this helper function and migrating this function logic
inside the SP_TABLES procedure whose language is pltsql and Like operator in pltsql supports square bracket wild card.

Task: BABEL-4128
Signed-off-by: Rohit Bhagat <rohitbgt@amazon.com>

Fix date functions to take into account the timezone setting (#2947)

Issues -

When the timezone parameter group setting is set to something other than the default UTC, Babelfish date
related functions like SYSDATETIME(), SYSDATETIMEOFFSET(), GETDATE() and CURRENT_TIMESTAMP don't take it
into account, while they should. The same issue occurs with a session-level timezone parameter.
SYSDATETIMEOFFSET() function returns datetimeoffset which does not store timezone due to incorrect casting
function.

Changes made to fix the issues -

Added appropriate CAST from TIMESTAMP to respective return types of date functions to ensure proper accounting for
timezone setting. Added timestamptz_datetimeoffset function for converting timestamp with time zone to
datetimeoffset. It ensures to store timezone property similar to timestamptz_datetime2 and timestamptz_datetime
functions.

Task: BABEL-5069
Signed-off-by: Anikait Agrawal <agraani@amazon.com>

refactoring (#2951)

Reduce the number of arguments passed to gen_createdb_subcmds & gen_dropdb_subcmds. Create a helper functions to generate RoleSpec & AccessPriv nodes from cstring. Create a new function for inserting fixed roles/users to bbf_authid_user_ext catalog during create database. Remove unused declaration update_RevokeRoleStmt

Issues Resolved: [BABEL-5274]

Signed-off-by: Tanzeel Khan tzlkhan@amazon.com

Support Cross-database references in views (#2899)

Support execution of views which references objects (tables/views/functions) from
across the databases. Here we are talking about Babelfish logical database (T-SQL database)
which is different from a physical Postgres database.
To support this, perform permission checks for cross database objects using session user (login)
instead of current_user (user of current T-SQL database). The reason login can be used for
permission check is since login is member of all it’s users, so it inherits all their permissions so it
will be able execute any cross database objects owned by its users.

This commit handles functions and tables/views separately for cross database permission
checks. For functions/procedures, a new hook `ExecFuncProc_AclCheck_hook` and for tables/views
existing `ExecutorStart_hook` will be used to decide whether to use session user or current_user
for permission check depending upon whether the object is from same or different database.
We will be using `is_schema_from_db` function to identify if the object is from different database
which performs a lookup into `babelfish_namespace_ext` catalog table which can be expensive as will
be doing it pretty frequently. So, added this table into SYSCACHE for better performance.
Tables/views permissions are handled slightly different than functions as we do not blindly want to check
the permissions against session user (current login) since permissions of RTEs inside a view are checked
against that view's owner which can very well be a user of some different database. So if we blindly check
permission against session user instead of view's owner then it would break view's ownership chaining.
Instead, we will replace `checkAsUser` with it's corresponding mapped login if present and only in cases
where `checkAsUser` is not set, we will replace it with session user (login). We are using login to allow cross
database queries since login can access all its objects across the databases. Getting mapped login to a
user require lookup into sys.babelfish_authid_user_ext catalog table using its primary key column (rolname) so
added this table is also into SYSCACHE.

Additionally, remove previous code to globally set current user to session user since newer logic
takes care of the permission check now.

Task: BABEL-5206
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/434

Revert unintended rebase commits "Support Cross-database references in views (#2899)"

This reverts commit 90a0b1fad4b49ce5e0523acad6ccbd0639e4f29a.

Revert unintended changes "refactoring (#2951)"

This reverts commit 37ce1309191dda4580d47ddc4f0aa43c0caa5ed0.

Revert unintended rebase commit "Fix date functions to take into account the timezone setting (#2947)"

This reverts commit 3df9fa08e8816c6cff6c7c5734950a5ea2e73142.

Revert unintended rebase commit "Fixed issue in SP_TABLES() when @table_name parameter has square brackets around underscore (#2940)"

This reverts commit 230c433b2ef9f20da8ab5fef226350e71a2461e1.

lejaokri commented on test/JDBC/expected/BABEL-5129-vu-verify.out:
```diff
@@ -0,0 +1,131 @@
+
+-- Test empty nvarchar variable
+DECLARE @nv nvarchar(10);
+SELECT @nv = '';
```
nice, thanks for adding this. can we try values that go beyond the limit, also NULL?

lejaokri replied:
Also, please verify the output with sql server instance.

manisha-deshpande replied:
> Also, please verify the output with sql server instance.

Tested queries with SQL Server 16 - ISNUMERIC results match PostgreSQL. Minor unrelated discrepancy in DATALENGTH noted (PG: 6, SQL Server: 5 for 12.3420000000) - Not sure if it's a PSQL/SQL internal storage behavior difference or a bug to be addressed. Can remove DATALENGTH from the select statements in tests if needed.

`SELECT ISNUMERIC(@a), LEN(@a), DATALENGTH(@a)
`

Commit:
Added null test case

Task: BABEL-5129
Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>

Commit:
Added boundary test cases

Signed-off-by: “manisha-deshpande” <“dmanisha.work@gmail.com”>
